usr/src/lib/libzfs_core/common/libzfs_core.c

```diff
@@ -239,6 +239,71 @@ lzc_promote(const char *fsname, char *snapnamebuf, int snapnamelen)
 	return (0);
 }
 
+int
+lzc_rename(const char *source, const char *target)
+{
+	zfs_cmd_t zc = { 0 };
+	int error;
+
+	(void) strlcpy(zc.zc_name, source, sizeof (zc.zc_name));
+	(void) strlcpy(zc.zc_value, target, sizeof (zc.zc_value));
+	error = ioctl(g_fd, ZFS_IOC_RENAME, &zc);
+	if (error != 0)
+		error = errno;
+	return (error);
+}
+
+int
+lzc_destroy(const char *fsname)
+{
+	int error;
+
+	nvlist_t *args = fnvlist_alloc();
+	error = lzc_ioctl(ZFS_IOC_DESTROY, fsname, args, NULL);
+	nvlist_free(args);
+	return (error);
+}
+
+/* XXX "inherit" received values (zfs inherit -S) ? */
+int
+lzc_inherit_prop(const char *fsname, const char *name)
+{
+	zfs_cmd_t zc = { 0 };
+	int error;
+
+	(void) strlcpy(zc.zc_name, fsname, sizeof (zc.zc_name));
+	(void) strlcpy(zc.zc_value, name, sizeof (zc.zc_value));
+	error = ioctl(g_fd, ZFS_IOC_INHERIT_PROP, &zc);
+	if (error != 0)
+		error = errno;
+	return (error);
+}
+
+int
+lzc_set_prop(const char *fsname, nvlist_t *props)
+{
+	nvpair_t *elem;
+	int error;
+
+	/* the list must have exactly one element */
+	if ((elem = nvlist_next_nvpair(props, NULL)) == NULL ||
+	    nvlist_next_nvpair(props, elem) != NULL)
```

Comment:
why do we have this restriction?  The ioctl handles multiple props.

Reply:
Although, we also probably don't want to "bless" this property interface either.  Delphix has some code for setting properties with channel programs but we haven't finished it yet :-(

Reply:
I don't recall the details now, but I think that there were some concerns about `ZFS_IOC_SET_PROP` being non-atomic when given multiple properties.
As the comment near `zfs_set_prop_nvlist` says:
```
 * This function is best effort. If it fails to set any of the given properties,
 * it continues to set as many as it can and returns the last error
 * encountered. If the caller provides a non-NULL errlist, it will be filled in
 * with the list of names of all the properties that failed along with the
 * corresponding error numbers.
```
But for some reason we wanted something that does all or nothing, so we opted to have a single-property interface.